smart_home.cpp

```diff
@@ -0,0 +1,92 @@
+#include "smart_home.h"
+
+void SmartHome::GetDataFromXml(const char* filename) {
+    tinyxml2::XMLDocument xmlDoc;
+    tinyxml2::XMLError err = xmlDoc.LoadFile(filename);
+    if (err != tinyxml2::XML_SUCCESS) {
+        throw std::invalid_argument("Error loading file!");
+    }
+
+    tinyxml2::XMLNode* pRoot = xmlDoc.RootElement();
+    if (pRoot == nullptr) {
+        throw std::invalid_argument("Invalid XML format!");
+    }
+
+    tinyxml2::XMLElement* pRoomElement = pRoot->FirstChildElement();
+    
+    if (pRoomElement == nullptr) {
+        throw std::invalid_argument("Room element is null!");
+    }
+
+    for (pRoomElement; pRoomElement != nullptr; pRoomElement = pRoomElement->NextSiblingElement("Room")) {
+        const char* roomName = pRoomElement->Attribute("name");
+
+        Room room(roomName);
+
+        for (tinyxml2::XMLElement* sensorElement = pRoomElement->FirstChildElement("Sensor"); sensorElement != nullptr; sensorElement = sensorElement->NextSiblingElement("Sensor")) {
```

Comment:
You should make your functions smaller by splitting into multiple small functions.
For example, you could have 2 (private) small functions, AddSensor/AddDevice that take the  xml element as param. It would improve readibility.

Reply:
Correct, I will update this as well